File: web/test/cli.test.ts
import * as fs from "node:fs";
import { beforeEach, describe, expect, test, vi } from "vitest";
import type { Mock } from "vitest";
import { getRuleset, parseArgs } from "../src/bin/index";
import parse from "../src/parse";
import renderHtml from "../src/renderHtml";

vi.mock("node:fs", () => ({
	readFileSync: vi.fn(),
	writeFileSync: vi.fn(),
}));

function resetMocks(): void {
	vi.clearAllMocks();
}

describe("CLI parseArgs", () => {
	test("parses basic input file", () => {
		resetMocks();
		const args = ["input.md"];
		const result = parseArgs(args);
		expect(result).toEqual({
			input: "input.md",
			output: null,
			ruleset: "extended",
			format: "html",
		});
	});

	test("parses input file with output", () => {
		resetMocks();
		const args = ["input.md", "--output", "output.html"];
		const result = parseArgs(args);
		expect(result).toEqual({
			input: "input.md",
			output: "output.html",
			ruleset: "extended",
			format: "html",
		});
	});

	test("parses input file with ruleset core", () => {
		resetMocks();
		const args = ["input.md", "--ruleset", "core"];
		const result = parseArgs(args);
		expect(result).toEqual({ input: "input.md", output: null, ruleset: "core", format: "html" });
	});

	test("parses input file with ruleset gfm", () => {
		resetMocks();
		const args = ["input.md", "--ruleset", "gfm"];
		const result = parseArgs(args);
		expect(result).toEqual({ input: "input.md", output: null, ruleset: "gfm", format: "html" });
	});

	test("parses input file with ruleset extended", () => {
		resetMocks();
		const args = ["input.md", "--ruleset", "extended"];
		const result = parseArgs(args);
		expect(result).toEqual({
			input: "input.md",
			output: null,
			ruleset: "extended",
			format: "html",
		});
	});

	test("parses all options", () => {
		resetMocks();
		const args = ["input.md", "--output", "output.html", "--ruleset", "gfm"];
		const result = parseArgs(args);
		expect(result).toEqual({
			input: "input.md",
			output: "output.html",
			ruleset: "gfm",
			format: "html",
		});
	});

	test("parses options in different order", () => {
		resetMocks();
		const args = ["--ruleset", "core", "input.md", "--output", "out.html"];
		const result = parseArgs(args);
		expect(result).toEqual({
			input: "input.md",
			output: "out.html",
			ruleset: "core",
			format: "html",
		});
	});

	test("parses -o shortcut for output", () => {
		resetMocks();
		const args = ["input.md", "-o", "output.html"];
		const result = parseArgs(args);
		expect(result).toEqual({
			input: "input.md",
			output: "output.html",
			ruleset: "extended",
			format: "html",
		});
	});

	test("parses -r shortcut for ruleset", () => {
		resetMocks();
		const args = ["input.md", "-r", "gfm"];
		const result = parseArgs(args);
		expect(result).toEqual({ input: "input.md", output: null, ruleset: "gfm", format: "html" });
	});

	test("parses both shortcuts -o and -r", () => {
		resetMocks();
		const args = ["input.md", "-o", "out.html", "-r", "core"];
		const result = parseArgs(args);
		expect(result).toEqual({
			input: "input.md",
			output: "out.html",
			ruleset: "core",
			format: "html",
		});
	});

	test("parses mixed shortcuts and long options", () => {
		resetMocks();
		const args = ["-r", "gfm", "input.md", "--output", "out.html"];
		const result = parseArgs(args);
		expect(result).toEqual({
			input: "input.md",
			output: "out.html",
			ruleset: "gfm",
			format: "html",
		});
	});
});

describe("CLI parseArgs error handling", () => {
	let mockConsoleError: Mock;

	beforeEach(() => {
		resetMocks();
		vi.spyOn(globalThis.process, "exit").mockImplementation(() => {
			throw new Error("process.exit called");
		}) as Mock;
		mockConsoleError = vi.spyOn(console, "error").mockImplementation(() => {}) as Mock;
	});

	test("exits with error when no input file", () => {
		expect(() => parseArgs([])).toThrow("process.exit called");
		expect(mockConsoleError).toHaveBeenCalledWith("Error: No input file specified");
	});

	test("exits with error when multiple input files", () => {
		expect(() => parseArgs(["input1.md", "input2.md"])).toThrow("process.exit called");
		expect(mockConsoleError).toHaveBeenCalledWith(expect.stringContaining("Multiple input files"));
	});

	test("exits with error when --output has no value", () => {
		expect(() => parseArgs(["input.md", "--output"])).toThrow("process.exit called");
		expect(mockConsoleError).toHaveBeenCalledWith("Error: --output requires a file path");
	});

	test("exits with error when -o has no value", () => {
		expect(() => parseArgs(["input.md", "-o"])).toThrow("process.exit called");
		expect(mockConsoleError).toHaveBeenCalledWith("Error: -o requires a file path");
	});

	test("exits with error when --ruleset has no value", () => {
		expect(() => parseArgs(["input.md", "--ruleset"])).toThrow("process.exit called");
		expect(mockConsoleError).toHaveBeenCalledWith(
			"Error: --ruleset requires a value (core, gfm, or extended)",
		);
	});

	test("exits with error when -r has no value", () => {
		expect(() => parseArgs(["input.md", "-r"])).toThrow("process.exit called");
		expect(mockConsoleError).toHaveBeenCalledWith(
			"Error: -r requires a value (core, gfm, or extended)",
		);
	});

	test("exits with error when -r has invalid ruleset", () => {
		expect(() => parseArgs(["input.md", "-r", "invalid"])).toThrow("process.exit called");
		expect(mockConsoleError).toHaveBeenCalledWith(expect.stringContaining("Invalid ruleset"));
	});

	test("exits with error when invalid ruleset", () => {
		expect(() => parseArgs(["input.md", "--ruleset", "invalid"])).toThrow("process.exit called");
		expect(mockConsoleError).toHaveBeenCalledWith(expect.stringContaining("Invalid ruleset"));
	});

	test("exits successfully on --help", () => {
		expect(() => parseArgs(["--help"])).toThrow("process.exit called");
	});

	test("exits successfully on -h", () => {
		expect(() => parseArgs(["-h"])).toThrow("process.exit called");
	});

	test("exits with error on unknown long option", () => {
		expect(() => parseArgs(["input.md", "--unknown"])).toThrow("process.exit called");
		expect(mockConsoleError).toHaveBeenCalledWith(expect.stringContaining("Unknown option"));
	});

	test("exits with error on unknown short option", () => {
		expect(() => parseArgs(["input.md", "-x"])).toThrow("process.exit called");
		expect(mockConsoleError).toHaveBeenCalledWith(expect.stringContaining("Unknown option '-x'"));
	});

	test("exits with error on unknown short option -a", () => {
		expect(() => parseArgs(["input.md", "-a"])).toThrow("process.exit called");
		expect(mockConsoleError).toHaveBeenCalledWith(expect.stringContaining("Unknown option '-a'"));
	});
});

describe("CLI main integration", () => {
	beforeEach(() => {
		resetMocks();
	});

	test("parses markdown to HTML with core ruleset", () => {
		const ruleset = getRuleset("core");
		const markdown = "# Heading\n\nParagraph text";
		const document = parse(markdown, ruleset, false);
		const html = renderHtml(document);
		expect(html).toContain("<h1>Heading</h1>");
		expect(html).toContain("<p>Paragraph text</p>");
	});

	test("parses markdown to HTML with gfm ruleset", () => {
		const ruleset = getRuleset("gfm");
		const markdown = "# Heading\n\n- List item 1\n- List item 2";
		const document = parse(markdown, ruleset, false);
		const html = renderHtml(document);
		expect(html).toContain("<h1>Heading</h1>");
		expect(html).toContain("<ul>");
	});

	test("parses markdown to HTML with extended ruleset", () => {
		const ruleset = getRuleset("extended");
		const markdown = "# Heading\n\n**bold** text";
		const document = parse(markdown, ruleset, false);
		const html = renderHtml(document);
		expect(html).toContain("<h1>Heading</h1>");
		expect(html).toContain("<strong>bold</strong>");
	});

	test("writes output to file when output is specified", () => {
		const html = "<h1>Test</h1>";
		fs.writeFileSync("output.html", html, "utf-8");
		expect(fs.writeFileSync).toHaveBeenCalledWith("output.html", html, "utf-8");
	});
});

describe("CLI getRuleset", () => {
	test("returns core ruleset", () => {
		const ruleset = getRuleset("core");
		expect(ruleset).toBeDefined();
	});

	test("returns gfm ruleset", () => {
		const ruleset = getRuleset("gfm");
		expect(ruleset).toBeDefined();
	});

	test("returns extended ruleset", () => {
		const ruleset = getRuleset("extended");
		expect(ruleset).toBeDefined();
	});
});

File: zig/test/cli_test.zig
const std = @import("std");

const allmark = @import("allmark");

const Ruleset = enum {
    core,
    gfm,
    extended,
};

const Args = struct {
    allocator: std.mem.Allocator,
    input: []const u8,
    output: ?[]const u8,
    ruleset: Ruleset,

    pub fn deinit(self: *const Args) void {
        if (self.output) |out| {
            self.allocator.free(out);
        }
    }
};

fn printUsage() void {
    std.debug.print("Usage: allmark <input-file> [--output <file>] [--ruleset <core|gfm|extended>]\n", .{});
    std.debug.print("  input-file  Path to the markdown file to convert\n", .{});
    std.debug.print("  --output     Path to output HTML file (optional, prints to stdout by default)\n", .{});
    std.debug.print("  --ruleset    Ruleset to use: core, gfm, or extended (default: extended)\n", .{});
}

fn parseArgs(allocator: std.mem.Allocator, args: [][]const u8) !Args {
    var result = Args{
        .allocator = allocator,
        .input = "",
        .output = null,
        .ruleset = .extended,
    };

    var i: usize = 0;
    while (i < args.len) {
        const arg = args[i];

        if (std.mem.eql(u8, arg, "--output")) {
            if (i + 1 >= args.len) {
                std.debug.print("Error: --output requires a file path\n", .{});
                printUsage();
                return error.MissingArgument;
            }
            result.output = try allocator.dupe(u8, args[i + 1]);
            i += 2;
        } else if (std.mem.eql(u8, arg, "--ruleset")) {
            if (i + 1 >= args.len) {
                std.debug.print("Error: --ruleset requires a value (core, gfm, or extended)\n", .{});
                printUsage();
                return error.MissingArgument;
            }
            const ruleset_str = args[i + 1];
            if (std.mem.eql(u8, ruleset_str, "core")) {
                result.ruleset = .core;
            } else if (std.mem.eql(u8, ruleset_str, "gfm")) {
                result.ruleset = .gfm;
            } else if (std.mem.eql(u8, ruleset_str, "extended")) {
                result.ruleset = .extended;
            } else {
                std.debug.print("Error: Invalid ruleset '{s}'. Must be core, gfm, or extended\n", .{ruleset_str});
                printUsage();
                return error.InvalidRuleset;
            }
            i += 2;
        } else if (std.mem.eql(u8, arg, "--help") or std.mem.eql(u8, arg, "-h")) {
            printUsage();
            std.process.exit(0);
        } else if (std.mem.startsWith(u8, arg, "--")) {
            std.debug.print("Error: Unknown option '{s}'\n", .{arg});
            printUsage();
            return error.UnknownOption;
        } else {
            if (result.input.len > 0) {
                std.debug.print("Error: Multiple input files specified: '{s}' and '{s}'\n", .{ result.input, arg });
                printUsage();
                return error.MultipleInputFiles;
            }
            result.input = arg;
            i += 1;
        }
    }

    if (result.input.len == 0) {
        std.debug.print("Error: No input file specified\n", .{});
        printUsage();
        return error.NoInputFile;
    }

    return result;
}

test "parseArgs basic input file" {
    const allocator = std.testing.allocator;
    const args = &[_][]const u8{"input.md"};
    const result = try parseArgs(allocator, args);
    defer result.deinit();

    try std.testing.expectEqualStrings("input.md", result.input);
    try std.testing.expect(result.output == null);
    try std.testing.expectEqual(Ruleset.extended, result.ruleset);
}

test "parseArgs with output file" {
    const allocator = std.testing.allocator;
    const args = &[_][]const u8{ "input.md", "--output", "output.html" };
    const result = try parseArgs(allocator, args);
    defer result.deinit();

    try std.testing.expectEqualStrings("input.md", result.input);
    try std.testing.expectEqualStrings("output.html", result.output.?);
    try std.testing.expectEqual(Ruleset.extended, result.ruleset);
}

test "parseArgs with ruleset core" {
    const allocator = std.testing.allocator;
    const args = &[_][]const u8{ "input.md", "--ruleset", "core" };
    const result = try parseArgs(allocator, args);
    defer result.deinit();

    try std.testing.expectEqualStrings("input.md", result.input);
    try std.testing.expect(result.output == null);
    try std.testing.expectEqual(Ruleset.core, result.ruleset);
}

test "parseArgs with ruleset gfm" {
    const allocator = std.testing.allocator;
    const args = &[_][]const u8{ "input.md", "--ruleset", "gfm" };
    const result = try parseArgs(allocator, args);
    defer result.deinit();

    try std.testing.expectEqualStrings("input.md", result.input);
    try std.testing.expect(result.output == null);
    try std.testing.expectEqual(Ruleset.gfm, result.ruleset);
}

test "parseArgs with ruleset extended" {
    const allocator = std.testing.allocator;
    const args = &[_][]const u8{ "input.md", "--ruleset", "extended" };
    const result = try parseArgs(allocator, args);
    defer result.deinit();

    try std.testing.expectEqualStrings("input.md", result.input);
    try std.testing.expect(result.output == null);
    try std.testing.expectEqual(Ruleset.extended, result.ruleset);
}

test "parseArgs with all options" {
    const allocator = std.testing.allocator;
    const args = &[_][]const u8{ "input.md", "--output", "output.html", "--ruleset", "gfm" };
    const result = try parseArgs(allocator, args);
    defer result.deinit();

    try std.testing.expectEqualStrings("input.md", result.input);
    try std.testing.expectEqualStrings("output.html", result.output.?);
    try std.testing.expectEqual(Ruleset.gfm, result.ruleset);
}

test "parseArgs options in different order" {
    const allocator = std.testing.allocator;
    const args = &[_][]const u8{ "--ruleset", "core", "input.md", "--output", "out.html" };
    const result = try parseArgs(allocator, args);
    defer result.deinit();

    try std.testing.expectEqualStrings("input.md", result.input);
    try std.testing.expectEqualStrings("out.html", result.output.?);
    try std.testing.expectEqual(Ruleset.core, result.ruleset);
}

test "parseArgs error - no input file" {
    const allocator = std.testing.allocator;
    const args = &[_][]const u8{};
    const result = parseArgs(allocator, args);

    try std.testing.expectError(error.NoInputFile, result);
}

test "parseArgs error - multiple input files" {
    const allocator = std.testing.allocator;
    const args = &[_][]const u8{ "input1.md", "input2.md" };
    const result = parseArgs(allocator, args);

    try std.testing.expectError(error.MultipleInputFiles, result);
}

test "parseArgs error - --output without value" {
    const allocator = std.testing.allocator;
    const args = &[_][]const u8{ "input.md", "--output" };
    const result = parseArgs(allocator, args);

    try std.testing.expectError(error.MissingArgument, result);
}

test "parseArgs error - --ruleset without value" {
    const allocator = std.testing.allocator;
    const args = &[_][]const u8{ "input.md", "--ruleset" };
    const result = parseArgs(allocator, args);

    try std.testing.expectError(error.MissingArgument, result);
}

test "parseArgs error - invalid ruleset" {
    const allocator = std.testing.allocator;
    const args = &[_][]const u8{ "input.md", "--ruleset", "invalid" };
    const result = parseArgs(allocator, args);

    try std.testing.expectError(error.InvalidRuleset, result);
}

test "parseArgs error - unknown option" {
    const allocator = std.testing.allocator;
    const args = &[_][]const u8{ "input.md", "--unknown" };
    const result = parseArgs(allocator, args);

    try std.testing.expectError(error.UnknownOption, result);
}

test "CLI integration - parse markdown with core ruleset" {
    const allocator = std.testing.allocator;

    const markdown = "# Heading\n\nParagraph text";
    var ruleset = try allmark.core.init(allocator);
    defer ruleset.blocks.deinit();
    defer ruleset.inlines.deinit();
    defer ruleset.renderers.deinit();

    const doc = try allmark.parse.execute(allocator, markdown, ruleset);
    defer doc.deinit(allocator);

    const html = try allmark.render(allocator, doc, null, false, null);
    defer allocator.free(html);

    try std.testing.expect(html.len > 0);
    try std.testing.expect(std.mem.indexOf(u8, html, "<h1>") != null);
    try std.testing.expect(std.mem.indexOf(u8, html, "Heading") != null);
}

test "CLI integration - parse markdown with gfm ruleset" {
    const allocator = std.testing.allocator;

    const markdown = "# Heading\n\n- List item 1\n- List item 2";
    var ruleset = try allmark.gfm.init(allocator);
    defer ruleset.blocks.deinit();
    defer ruleset.inlines.deinit();
    defer ruleset.renderers.deinit();

    const doc = try allmark.parse.execute(allocator, markdown, ruleset);
    defer doc.deinit(allocator);

    const html = try allmark.render(allocator, doc, null, false, null);
    defer allocator.free(html);

    try std.testing.expect(html.len > 0);
    try std.testing.expect(std.mem.indexOf(u8, html, "<h1>") != null);
    try std.testing.expect(std.mem.indexOf(u8, html, "<ul>") != null);
}

test "CLI integration - parse markdown with extended ruleset" {
    const allocator = std.testing.allocator;

    const markdown = "# Heading\n\n**bold** text";
    var ruleset = try allmark.extended.init(allocator);
    defer ruleset.blocks.deinit();
    defer ruleset.inlines.deinit();
    defer ruleset.renderers.deinit();

    const doc = try allmark.parse.execute(allocator, markdown, ruleset);
    defer doc.deinit(allocator);

    const html = try allmark.render(allocator, doc, null, false, null);
    defer allocator.free(html);

    try std.testing.expect(html.len > 0);
    try std.testing.expect(std.mem.indexOf(u8, html, "<h1>") != null);
    try std.testing.expect(std.mem.indexOf(u8, html, "<strong>") != null);
}

test "CLI integration - write output to file" {
    const allocator = std.testing.allocator;

    const markdown = "# Test";
    var ruleset = try allmark.core.init(allocator);
    defer ruleset.blocks.deinit();
    defer ruleset.inlines.deinit();
    defer ruleset.renderers.deinit();

    const doc = try allmark.parse.execute(allocator, markdown, ruleset);
    defer doc.deinit(allocator);

    const html = try allmark.render(allocator, doc, null, false, null);
    defer allocator.free(html);

    try std.testing.expect(html.len > 0);
    try std.testing.expect(std.mem.indexOf(u8, html, "<h1>Test</h1>") != null);
}
